### ISSUE TITLE
[TRACKER] Porting the demos to Godot 4

### ISSUE THREAD

**Issue by aaronfranke** [godotengine/godot-demo-projects#697]
This issue tracks the progress of porting the Godot demo projects to the current master branch of Godot Engine (what will eventually be Godot 4.0). All of this work is done on the [`4.0-dev`](https://github.com/godotengine/godot-demo-projects/tree/4.0-dev) branch as to not confuse 3.x users who want to download the demos. The `4.0-dev` branch will be merged into [`master`](https://github.com/godotengine/godot-demo-projects/tree/master) in the future, likely in the coming weeks. This list is in alphabetical order by folder path.

* ✅ and a checked box means that the demo is fully ported and working with absolutely 0 bugs. Even a false warning will prevent a demo from getting a ✅.
* 🚧 means that the demo has only had [the script](https://gist.github.com/aaronfranke/79b424226475d277d0035b7835b09c5f) run on it and still needs to be ported manually.
* 🐛 means that there are engine (not project) bugs that show up in this project. Projects that have bugs but the cause is unknown don't get 🐛.
* 🐞 means that there are minor bugs such as errors printing in the console.
* 🌋 means that there is a bug with rendering (possibly a problem with the project).
* 📜 means that there is a bug with scripting (can be GDScript/C#/VisualScript).
* 🖽 means that this project uses TileMap. 3.x tile data doesn't work in master.
* 💥 means that opening the editor with this project causes the engine to crash.

- [ ] 🚧⚛️ Bullet Shower
- [ ] 🚧🌋 Dodge the Creeps
- [ ] 🚧📜 Hierarchical Finite State Machine
- [ ] 🚧📜 GD Paint
- [ ] 🚧🌋 HDR for 2D
- [ ] 🚧🖽 Hexagonal Game
- [ ] 🚧 Scene Instancing Demo
- [ ] 🚧🖽 Isometric Game
- [ ] 🚧🖽 Kinematic Character 2D
- [ ] 🚧 2D Lights as Mask
- [ ] 🚧 2D Lights and Shadows
- [ ] 🚧 Navigation Polygon 2D
- [ ] 🚧🖽 Grid-based Pathfinding with Astar
- [ ] 🚧 2D Particles
- [ ] 🚧🖽 Physics-Based Platformer 2D
- [ ] 🚧📜 2D Physics Tests
- [ ] 🚧🖽 Platformer 2D
- [ ] 🚧 Pong with GDScript
- [ ] 🚧🖽 JRPG Demo
- [ ] 🚧 Screen Space Shaders
- [ ] 🚧 Skeleton2D Demo
- [ ] 🚧 2D Shaders for Sprites
- [ ] 🚧 Tween Demo
- [ ] 🚧 3D Global Illumination
- [ ] 🐛💥 3D Inverse Kinematics
- [x] ✅ Kinematic Character 3D
- [x] ✅ Material Testers
- [ ] 🚧 3D Navigation Mesh
- [ ] 🐛🐞 3D Physics Tests
- [ ] 🚧 Platformer 3D
- [x] ✅ RigidBody Character 3D
- [ ] 🐞 Truck Town
- [ ] 🚧 Volumetric Fog
- [ ] 🌋📜 Voxel Game
- [ ] 🚧📜 3D Waypoints
- [ ] 🐞 BPM Sync Demo
- [x] ✅ Audio Device Changer Demo
- [ ] 🚧 Audio Generator Demo
- [x] ✅ Audio Mic Record Demo
- [x] ✅ Audio MIDI Piano Demo
- [x] ✅ Audio Spectrum Demo
- [ ] 🚧 BIDI and Font Features
- [ ] 🚧 Control Gallery
- [ ] 🚧 Drag and Drop (GUI)
- [ ] 🚧 Input Mapping GUI
- [ ] 🚧 Pseudolocalization Demo
- [ ] 🚧 RegEx (Regular Expressions)
- [ ] 🚧 Rich Text Label with BBCode
- [ ] 🚧 Signed Distance Field Font
- [ ] 🚧 GUI Theming Override Demo
- [ ] 🚧 Translation Demo
- [ ] 🚧 UI Mirroring Demo
- [ ] 🚧 Autoload (Singletons)
- [ ] 🚧 Background Thread Loading Demo
- [ ] 🚧 Multiple threads loading
- [ ] 🚧 Scene Changer
- [ ] 🚧 Saving and Loading (Serialization)
- [ ] 🚧 Loading in a Thread
- [ ] 🚧 2.5D Demo with GDScript
- [ ] 🚧 Joypads Demo
- [ ] 🚧 Matrix Transform
- [ ] 🚧 OpenSimplexNoise Viewer
- [ ] 🚧 Operating System Testing
- [ ] 🚧 Pause Demo
- [ ] 🚧 Window Management
- [ ] 🚧 Android In-App Purchases
- [ ] 🚧 Interactive Cubes
- [ ] 🚧 Multitouch View
- [ ] 🚧 Accelerometer, Magnetometer and Gyro Demo
- [ ] 🚧 2.5D Demo with C#
- [ ] 🚧 Android in-app purchases with C#
- [ ] 🚧 Dodge the Creeps with C#
- [ ] 🚧 Pong Multiplayer with C#
- [ ] 🚧 Pong with C#
- [ ] 🚧 Multiplayer Bomber
- [ ] 🚧 Pong Multiplayer
- [ ] 🚧 WebRTC Minimal Connection
- [ ] 🚧 WebRTC Signaling Example
- [ ] 🚧 WebSocket Chat Demo
- [ ] 🚧 WebSocket Minimal Demo
- [ ] 🚧 WebSocket Multiplayer Demo
- [ ] 🚧 Plugin Demos
- [x] ✅ 2D in 3D Demo
- [ ] 🐞 3D in 2D Demo
- [x] ✅ 3D Viewport Scaling
- [x] ✅ Dynamic Split Screen
- [ ] 🐛🐞 GUI in 3D Demo
- [x] ✅ Screen Capture
- [ ] 🚧 Circle Pop
- [ ] 🚧 Multitouch View with VisualScript
- [ ] 🚧 Pong with VisualScript

**Comment by Calinou** [comment]
I have an ongoing 4.0 port of the 3D Global Illumination demo: https://github.com/Calinou/godot-demo-projects/tree/add-global-illumination-demo-2-4.0

**Comment by voylin** [comment]
I have fixed the Pause demo with pull request #701
And I checked the Autoload(Singletons) in pull request #702
I hope I'm helping in finishing off all these tasks. I'm pretty now to contributing so please tell me when I'm doing things wrong!

**Comment by GeorgeS2019** [comment]
@Calinou I wonder if it makes sense to discuss here what to do with the Godot4 c# demo examples. Should there be discussions among Godot4 mono users or we wait until Godot4 dotnet6 is merged into the master?

@voylin Do you do Godot4 c#? Thank you for volunteering.

**Comment by Calinou** [comment]
@GeorgeS2019 Porting C# demos will have to wait until the .NET 6 work is merged in `master`. Until this happens, the API is too unstable for porting to be worth it.

**Comment by aaronfranke** [comment]
@GeorgeS2019 That's a good point, I added a new emoji to this tracker to reflect this status ❌.

**Comment by voylin** [comment]
I've been trying to get some of the demo's fixed but 2D Lights as Mask and 2D Lights and Shadows are not possible yet. There are some rendering issues with light2D which make the demo's not work.

**Comment by Calinou** [comment]
> I've been trying to get some of the demo's fixed but 2D Lights as Mask and 2D Lights and Shadows are not possible yet. There are some rendering issues with light2D which make the demo's not work.

Indeed, see https://github.com/godotengine/godot/issues/55806.

**Comment by alex9099** [comment]
I just tested the voxel game and it seems to work without problems (no debugger errors nor anything seemingly wrong rendering wise) on alpha 7

**Comment by voylin** [comment]
> @voylin Do you do Godot4 c#? Thank you for volunteering.

@GeorgeS2019 Sorry for the late reply, hadn't seen the message. I am willing to test the c# stuff as well, but we will have to wait until the beta I guess.  Like Calinou said, .NET 6 stuff should first be merged.

**Comment by voylin** [comment]
2D Physics demo is having a bug which is occuring because of find_children. Issue godotengine/godot#54939 has to be solved first

**Comment by voylin** [comment]
> I just tested the voxel game and it seems to work without problems (no debugger errors nor anything seemingly wrong rendering wise) on alpha 7

@alex9099 The voxel demo still has some rendering bugs. @aaronfranke worked on this demo in #694
But when testing it I found rendering issues especially when breaking blocks in the flat space, not all the faces on the sides don't get updated. This demo will probably need to wait until the beta releases before it can be checked of the list.

**Comment by Anutrix** [comment]
All these need to be rechecked with Godot 4 beta or newer. Probably most of them are broken again now.

Tried Bullet Shower(tagged ✅) on master build and got: `Line 59:Function "update()" not found in base self.`
Built-in 3to4 Converter didn't help either.

**Comment by Calinou** [comment]
I'm going through the `4.0-dev` demos to update them as we speak.

**Comment by ScriptsEngineer** [comment]
**Multiplayer Pong**: Ported to godot 4 beta 5 -> https://github.com/godotengine/godot-demo-projects/pull/792

**Comment by zoeesilcock** [comment]
I have noticed that simply opening and running the demo projects on Godot 4.0b10 causes diffs in some scenes. It can be simple stuff like certain attributes have changed place or bigger stuff like removed attributes. Should these changes be committed as we find them or is there a plan to go through and commit all these scene-diffs caused by Godot 4.0 at one time later on?

**Comment by Calinou** [comment]
> I have noticed that simply opening and running the demo projects on Godot 4.0b10 causes diffs in some scenes. It can be simple stuff like certain attributes have changed place or bigger stuff like removed attributes. Should these changes be committed as we find them or is there a plan to go through and commit all these scene-diffs caused by Godot 4.0 at one time later on?

I think we can do this once 4.0 gets closer to release, as some bug fixes may still cause changes to occur in scene files.

**Comment by hsandt** [comment]
I tried to run Platformer 2D and it failed immediately due to API change in Game.gd. I could fix the pure script errors but the project is buggy after that (incorrect behavior). I suppose it's worth doing a first PR just for the scripting error fixing, but since I haven't forked the repo yet I'll just drop my notes here for now:

`OS.window_fullscreen = not OS.window_fullscreen`
becomes
`toggle_fullscreen()`
where the function is defined as:
```gd
func toggle_fullscreen():
	if DisplayServer.window_get_mode() not in \
			[DisplayServer.WINDOW_MODE_FULLSCREEN, DisplayServer.WINDOW_MODE_EXCLUSIVE_FULLSCREEN]:
		DisplayServer.window_set_mode(DisplayServer.WINDOW_MODE_FULLSCREEN)
	else:
		DisplayServer.window_set_mode(DisplayServer.WINDOW_MODE_WINDOWED)
```

`get_tree().set_input_as_handled()`
becomes
`get_viewport().set_input_as_handled()`

However, I noticed that both are mentioned in https://github.com/godotengine/godot/blob/master/editor/project_converter_3_to_4.cpp (with the subtlety that fullscreen has two variants and the converter only cares about exclusive fullscreen, so user would need a bit of manual work after the auto-conversion to get the exact behavior they want), which is itself used by command `godot4 --convert-3to4`.

I suspect that because only the gist converted script https://gist.github.com/aaronfranke/79b424226475d277d0035b7835b09c5f was run, the newest changes in project_converter_3_to_4.cpp have not been taken into account. So before I start forking the project and converting things manually, maybe we should try to run the latest --convert-3to4 on the projects that still need it to see how far we can automatically update things?

I think this is also a good idea to spot what --convert-3to4 can do and what it can't, and for anything that it didn't convert, we should decide if:
a. the missed conversion is simple enough to be integrated into project_converter_3_to_4.cpp
b. the missed conversion requires human interaction to decide subtle edge case and wanted behavior, in which case it should be added to some migration guide (like https://gist.github.com/WolfgangSenff/168cb0cbd486c8c9cd507f232165b976 but I think this one is unofficial, so we could do the same with an official one; on a personal note, this gist receives a lot of comments and seems to be a good basis to work from)

At the end, we would have a compilation of all the manual things we did to upgrade the demo projects that would contribute to this migration guide.

**Comment by aaronfranke** [comment]
@hsandt You are welcome to convert with any of these means, and submit any improvements you'd like to the converter. Just make sure that the final PR you submit has been hand-checked to have only good changes.

**Comment by hsandt** [comment]
For 2D demos using the tilemap:

In Platformer 2D, I noticed this tilemap break:

![image](https://user-images.githubusercontent.com/1067772/219716748-a22a564a-e207-48b7-bc20-4e7fe907c6b8.png)

due to Rendering Texture Origin being by default (0, 0) which means centered, causing tiles with Texture Region Size not matching TileSet Tile Size to have a representative area (with orange borders inside the green rectangle) to be misaligned.

I'm gonna fix the Texture Origin to some value so that the orange box matches the top of the tile, but it would be easier with some Texture Origin anchor direction I think (an enum with values: Center (current), Top-left, Top-middle, Top-right, Middle-left, Middle-right, Bottom-left, Bottom-middle, Bottom-right).

Or maybe the better way of doing this is to just split the big 32x64 tile into two parts of 32x32? I'm doing this in this in my own project (and also using auto-tile even with slopes, see https://github.com/godotengine/godot-proposals/discussions/6157).

For now I'll just fix the offset for minimal change, but I'd like to know the purpose of the demo to know better how I should use Godot 4's new features. Basically, are we demonstrating the tilemap too and trying to do things nicely in this demo, or are we just demonstrationg platforming, leaving TileSet demonstration to some other future TileMap demo?

**Comment by hsandt** [comment]
I opened a new branch for Demo Platformer 2D porting: https://github.com/hsandt/godot-demo-projects/tree/4.0-dev-platformer

I will send a PR when I get something decent to play. Unless you prefer that I regularly send small PR as long as there is some visible improvement?

**Comment by Calinou** [comment]
> I opened a new branch for Demo Platformer 2D porting: [hsandt/godot-demo-projects@`4.0-dev-platformer`](https://github.com/hsandt/godot-demo-projects/tree/4.0-dev-platformer?rgh-link-date=2023-02-17T20%3A04%3A04Z)
> 
> I will send a PR when I get something decent to play. Unless you prefer that I regularly send small PR as long as there is some visible improvement?

You could open a draft PR now, or wait later and open a full PR directly.

**Comment by voylin** [comment]
Maybe a good idea to remove the visual scripting ones as the folder has already been deleted from the branch and the visual scripting has been removed from the engine in 4.0
- Multitouch View with VisualScript
- Pong with VisualScript

Will this script be updated again once 4.0 is released? I will have some spare time at the end of this month and would be able to go over some of these projects to see if I can check/fix the demo projects for 4.0, would be helpful to see here which ones aren't finished yet.(probably quite some work to keep this list up to date but just asking to check)

**Comment by hsandt** [comment]
> You could open a draft PR now, or wait later and open a full PR directly. I personally prefer opening a draft PR as soon as I have something workable, so someone else can continue the work in case I no longer have time to complete it.

Here you go. BTW should I prefix all my commits with [4.0] instead of [3to4] ? I changed convention between my commits and my PR title (maybe it doesn't matter if you squash the whole thing, but it will if you preserve the full commit history).

**Comment by hsandt** [comment]
If you have trouble with the camera being too zoomed out in your demo, mind that the Zoom factor convention has been reversed: it now zooms in when the value increases, zooms out when the value decreases. So you have to reverse the value: in the case of 2D platformer, I changed (0.5, 0.5) -> (2.0, 2.0).



![image](https://user-images.githubusercontent.com/1067772/221283972-041d45e8-8d96-4421-9e89-ae4c7869cab0.png)

![image](https://user-images.githubusercontent.com/1067772/221284001-78804fa8-e066-416d-9109-36425aee5d46.png)

![image](https://user-images.githubusercontent.com/1067772/221284111-38ff6169-feaf-4157-a843-ca645deccfe6.png)

I suppose it could be added to --convert-3to4, but since the property name is not changing, it is particularly difficult to decide what to do, unless you are certain that you are actually converting a Godot 3 project. Basically, since the change is about reversing values, it is not idempotent, so you cannot use --convert-3to4 multiple times, guaranteeing no further changes.

I suppose we could add an option --force-convert-3to4 or something that will convert values this way (maybe only if project is actually using Godot 3; this won't work for people who started converting their project with an older version of the converter, unfortunately).

Or we could just show a warning when Camera zoom values != (1, 1) are detected (in scene or in script; in script, we could even add a comment above).

But if it's too complicated, we could add this to some upgrading guides so developer can just do it manually. For now I'll just comment on the two unofficial guides I know (https://gist.github.com/WolfgangSenff/168cb0cbd486c8c9cd507f232165b976 and https://www.reddit.com/r/godot/comments/tsxgnj/migrating_from_3x_to_40/).

**Comment by InfiniteProductions** [comment]
3D/MaterialTesters with 4.0 RC6 on my computer give me this:

![godot-4rc6-3D-mattesters-issue](https://user-images.githubusercontent.com/16732089/221657714-95d2dd1e-16a2-4d87-b1ad-5e2678cf1b77.png)

importing assets (_sand_rough.jpg_) stuck à 94%, CPU is 100% use (the decrease in CPU graph comes from a change of affinity from all to only 2/8 cores), then I decide to kill Godot.

Trying to reimport it directly from project list makes Godot stuck at "park.hdr" 0%, still 100% CPU usage, then finally project is loaded after several minutes:
![godot-4rc6-3D-mattesters-issue2](https://user-images.githubusercontent.com/16732089/221659725-1ec33a58-d243-45d3-9700-47907f06ab4f.png)

I don't call this working actually.

For the record, it's on an Asus G750 JH with a core i7 and a GeForce GTX 780M, I know it's far from being the best available these day, still it's not yet a low-end computer.
(tool used is System Informer)

**Comment by Calinou** [comment]
> Trying to reimport it directly from project list makes Godot stuck at "park.hdr" 0%, still 100% CPU usage, then finally project is loaded after several minutes:

This is likely due to https://github.com/godotengine/godot/issues/61948.

**Comment by InfiniteProductions** [comment]
I've checked all the import settings after the project is reopened in the editor, all are using VRAM compression, including HDR backgrounds, except one.

Only _sand_shine.jpg_ texture use lossless mode, it's 1024x1024 RGB8 texture, no mipmap using ~3MB memory, I don't believe it could be responsible for the import delay alone. And I don't know also if settings has been somehow modified automatically by the editor as I did not change anything not delete the _.godot_ folder inside the project.

By the way, opening the project again is as quick as it is expected.
And to answer your question, no, I'm using official 64bits build (v4.0.rc6.official [0cd148313]).

**Comment by InfiniteProductions** [comment]
I've done more testing on RC6 build:

platformer 3D failed (it looks like there is a regression), I just run the project, then I get this error right away, with a sounds looping while the debugger is running:
![godot-4rc6-3D-platformer-issue](https://user-images.githubusercontent.com/16732089/222113148-85a6b150-cae2-4725-9f54-27af5fc17084.png)

physicsTest (3D) also have at least one case of failure:

with tests/functional tests/collision pairs using collider type "capsule" and shape type "capsule":
![godot-4rc+-physicsTest-issue](https://user-images.githubusercontent.com/16732089/222113868-68a76a86-2063-461c-8106-91a47437077e.png)

![godot-4rc+-physicsTest-issue2](https://user-images.githubusercontent.com/16732089/222113905-68c6e69a-7aed-4593-9528-96b58a21ed88.png)

To reproduce the error, run the project, select the test "functional tests/collision pairs", select "test options" menu, click on "collider type" "capsule (3)" then "shape type" capsule checkbox => error occurred.
(it's happened for pretty much every choices, I use box, sphere, convex/concave polygon)
In fact, selecting the second option trigger the issue.

and there are a lot's of errors:
![godot-4rc+-physicsTest-errors-output](https://user-images.githubusercontent.com/16732089/222113927-9f5f5932-db60-47a7-8dd7-551a5b329792.png)

**Comment by hsandt** [comment]
So, how do we publish the demos already ported to Godot 4.0, now that it's out? I'm thinking in particular about 2D platformer which recently reached a decent state thanks to https://github.com/godotengine/godot-demo-projects/pull/851

**Comment by Calinou** [comment]
> Only sand_shine.jpg texture use lossless mode, it's 1024x1024 RGB8 texture, no mipmap using ~3MB memory, I don't believe it could be responsible for the import delay alone. And I don't know also if settings has been somehow modified automatically by the editor as I did not change anything not delete the .godot folder inside the project.

That texture isn't used anywhere in the project. It could likely be removed at this point.

**Comment by InfiniteProductions** [comment]
Platformer3D still fail with the same error (see my previous report above) with latest stable release (v4.0.stable.official [92bee43ad]).

**Comment by Calinou** [comment]
> Platformer3D still fail with the same error (see my previous report above) with latest stable release (v4.0.stable.official [92bee43ad]).

The demo needs updating to account for recent breaking changes in AnimationTree: https://github.com/godotengine/godot/pulls?q=is%3Apr+sort%3Aupdated-desc+transition_request+is%3Aclosed

**Comment by InfiniteProductions** [comment]
> The Platformer 3D demo needs updating to account for recent breaking changes in AnimationTree: https://github.com/godotengine/godot/pulls?q=is%3Apr+sort%3Aupdated-desc+transition_request+is%3Aclosed
> 
> Note that merely changing `/current` to `/transition_request` isn't enough, as it'll break the jumping animation and cause an error to be spammed every frame.

ok thanks.
The first post by @aaronfranke need to be updated accordingly as it shows the demo is fully working with zero issue (check mark + checkbox checked).

I guess we have too check all demos again on the stable release.

**Comment by aaronfranke** [comment]
Thanks for the reminder @InfiniteProductions. I have reset the tracker to have all boxes unchecked to ensure we re-check everything. Contributors such as @Calinou, please re-check anything that's known to be working.

**Comment by jgalbertus** [comment]
The Audio Generator Demo seems broken in Godot 4.0.  The buffer fills, but does not seem to delete as it plays.  The audio generator makes a boop, but then won't accept any more frames.  Any help would be much appreciated.

**Comment by InfiniteProductions** [comment]
@jgalbertus you may need to be more specific.

Tested with v4.0.1.stable.official [cacf49999] on Windows, frame count increase and decrease as tone is played, I just add some print statement with _to_fill_ value.

**Comment by GeorgeS2019** [comment]
- [ ] 3d/add-physical-light-camera-units-demo

**Comment by RPicster** [comment]
Hey :)
What would be the best way to help? The list at the top is still completely blank.
How is the process of porting projects to G4? Just open a PR?
Where will the G3 versions of the demos live?
Sorry if this has been answered or planned somewhere and I just ask the obvious questions...

**Comment by InfiniteProductions** [comment]
@RPicster the status has been reset a few weeks ago by @aaronfranke, I've tested again a bunch of demos with various 4.0 release along the way, there was quite a rush on releases and I couldn't catch up and redo all tests for every releases as it should be done, I've fill a file with all the results and status I've got, see my reply above from 3 weeks ago.

If you find an issue and/or have a fix, do not hesitate to fill a PR. You could also test the demos I didn't test or test those which have been checked first again with the last release.
I just report editor message from import and runtime warning/error + if it's work or not.

It looks like releases process is slowing down, it would leave us time to test all 89 or so on the very last release if there are enough people to test I guess.

**Comment by RPicster** [comment]
@InfiniteProductions Ok! Thanks for getting back to me.
Now how is support for the 3.x demos planned?
3.x is still LTS, so the demos should not be hidden in some branch or old commit from my pov. Would it make sense to handle them in separate repos?

**Comment by Calinou** [comment]
There's are `3.x` and `3.5` branches now. While we recommend focusing on the `master` branch, you can still create new demos for the `3.x` branch, and we can cherry-pick changes compatible with 3.5.x to the `3.5` branch.

**Comment by Anutrix** [comment]
Continuous 4.0.x releases is making demos never reach any stability.
**`Waiting for Godot 4`** has become **`Waiting for Godot 4 Demos`**.
Demos aren't few ranks behind Godot. They are few laps behind.

One drastic idea I can think off is temporary halting Godot's new 2-week release cycle and finishing as much demo upgrades as possible. And resume normal cycle once the demos are in passable state. Then, they can then be improved over time slowly.

Making we can even set aside some time for it once every quarter/semester/year for demo-upgrade.

Lack of minimalistic but ready demos for things like 'websocket-minimal' is really make it hard for Devs understand how to use them or understand the new changes. Porting manually becomes harder too.

**Comment by GeorgeS2019** [comment]
@Anutrix 

Godot is a community effort, FIRST, organize what is basic and missing in Godot4 and raise awareness in the community to get someone making it works in Godot4.

If you are not coding, keep hammering to make others to make the demo(s) happen.

**Comment by GeorgeS2019** [comment]
@Anutrix 
> Maybe we need to arrange the demo list from least complex to most complex first.

If you have an idea to share.  => Do it so the Godot management can choose your idea and integrate it.

**Comment by Calinou** [comment]
> One drastic idea I can think off is temporary halting Godot's new 2-week release cycle and finishing as much demo upgrades as possible. And resume normal cycle once the demos are in passable state. Then, they can then be improved over time slowly.
Making we can even set aside some time for it once every quarter/semester/year for demo-upgrade.

This is exactly what I did back in January :slightly_smiling_face:

**Comment by InfiniteProductions** [comment]
I believe some coordination would be more than welcome in order to made progress all together without stepping on each other foot :).
Maybe take charge of some testing (5, 10, or more demos) on a given version & architecture, report then have a regularly updated post like the tracker, but maybe something every willing participant can update via pull requests.

Issues could be created and assigned to people, but maybe it would need another project just for that, it may worth it although.

**Comment by aaronfranke** [comment]
@Anutrix I used to be on top of things, but I no longer have an abundance of free time like I used to, so I can't work on the demos. Someone needs to step up and actually do the work. It could be you, or someone you know. Or you could raise awareness on community channels.

There are a lot of people on places Discord that say they want to contribute, and then tell them that Godot needs people to contribute to the demos, but so far nobody has stepped up to actually do the work. I have even offered to pay people money to work on the demos and still nobody has stepped up.

**Comment by Anutrix** [comment]
One thing that stops me from updating official demos is what amount of changes is okay and what counts as stable. Do I need to it to be running in all 3 rendering drivers. Is running with few warnings are fine?. Should I test with .NET version of Godot too?, etc. I still don't know the answer to some of these questions.

I think many contributors gets discouraged because they don't know what should be basic tests/requirements when they make demo fixes or upgrades. In order words, 'Is my changes for 'Official Demos' good enough?'.

It's much harder than making changes in https://github.com/godotengine/godot code. There we just need fix the bug and check for stability before starting with draft MR.
Here, just opening a file will make changes. For example, on just opening `websocket_chat` in Godot 4.0.2, we see this diff:
```
$ git diff
diff --git a/networking/websocket_chat/project.godot b/networking/websocket_chat/project.godot
index 68cd0394..44581c17 100644
--- a/networking/websocket_chat/project.godot
+++ b/networking/websocket_chat/project.godot
@@ -8,22 +8,6 @@

 config_version=5

-_global_script_classes=[{
-"base": "Node",
-"class": &"WebSocketClient",
-"language": &"GDScript",
-"path": "res://websocket/WebSocketClient.gd"
-}, {
-"base": "Node",
-"class": &"WebSocketServer",
-"language": &"GDScript",
-"path": "res://websocket/WebSocketServer.gd"
-}]
-_global_script_class_icons={
-"WebSocketClient": "",
-"WebSocketServer": ""
-}
-
 [application]

 config/name="WebSocket Chat Demo"
```
This scares me as a contributor. Some FAQs in Official Docs saying 'Minor changes in project.godot when opening the project is expected' would be encouraging.

Btw. I was looking for demo guidelines.
Do we have any guidelines, FAQs or similar page for 'Contributing to Godot Demo Projects'? The reason my first contribution to a public open-source project was Godot Engine was because of its amazing 'https://docs.godotengine.org/en/latest/contributing/ways_to_contribute.html' section. It even helped me contribute to non-godot public projects.

Can we have something something similar for the `Godot Demo Projects`? We can start with a simple FAQ too.


Lastly, I'm interested in getting websocket demos working. I will make an MR if and when I succeed. It will most probably be rough but I guess it can be fixed on first review by maintainer. If all goes well, I will try making more MRs.

**Comment by GeorgeS2019** [comment]
@Anutrix 

I would suggest start with what u are comfortable , which render environment and share openly what u are not comfortable so others with the right competency which can complement u to work on the missing part.

The key is communication 

Start the communication so others know how to support u moving forwards.

**Comment by Calinou** [comment]
> . Do I need to it to be running in all 3 rendering drivers. Is running with few warnings are fine?. Should I test with .NET version of Godot too?, etc. I still don't know the answer to some of these questions.

You only need to test the rendering method the project currently uses on desktop. You don't need to test with a .NET version of Godot unless the project uses C#.

**Comment by InfiniteProductions** [comment]
I'm also willing to help, I just need a way to know who test what with what results with a good enough update frequency, even if it's means we update it ourselves, to go on the right path.

**Comment by hsandt** [comment]
For 2D Lights as Mask, I found that setting Burano TextureRect > CanvasItemMaterial > Light Mode from Light Only to Normal makes the photo appears as it should, with the textured lights:

![image](https://github.com/godotengine/godot-demo-projects/assets/1067772/92e76eb8-49b1-464d-bfef-688a14ef71f2)

![image](https://github.com/godotengine/godot-demo-projects/assets/1067772/ee61d4f5-af91-4f06-b586-f6f7bfd5d7ff)

**Comment by hsandt** [comment]
So I tested the demo in Godot 3 to compare, and Light Only works, and the result is different:

![image](https://github.com/godotengine/godot-demo-projects/assets/1067772/f5323a6a-3ab4-47a8-9854-ca46b0a32530)

and that's the intended result. So switching to Normal is just hiding the problem.

In the meantime I thought the core issue for Light Only failing: https://github.com/godotengine/godot/issues/44559

**Comment by Calinou** [comment]
Several demos in 3.x were revamped with new art assets:

- https://github.com/godotengine/godot-demo-projects/pull/898
- https://github.com/godotengine/godot-demo-projects/pull/904
- https://github.com/godotengine/godot-demo-projects/pull/905
- https://github.com/godotengine/godot-demo-projects/pull/915

Remember to use the commits linked at the end of each PR as a base, as they include some additional changes that aren't in the PRs themselves.

These demos should be ported to `master` again, since a lot of things have changed within them. It'll be faster to discard the `master` copy and start converting the current `3.x` demo from scratch.

**Comment by GeorgeS2019** [comment]
@Calinou 

Today is an important 4.1 milestone. Wonder if there are ways to track demo reflecting new features or changes in 4.1

- [ ] Rendering
      - [ ] NoiseTexture3D 
- [ ] Navigation
     - [ ] [Support different actor types ](https://docs.godotengine.org/en/stable/tutorials/navigation/navigation_different_actor_types.html)

**Comment by TheYellowArchitect** [comment]
Hello! I see that once I download 4.1 from the official website, there are 2 great links below the donations button. The first redirects to the docs, the 2nd to the asset library demos. All demos seem 3.5. Until most demos are updated, I suggest hiding the 2nd button.

**Comment by TheYellowArchitect** [comment]
I plan to update all demos under loading folder, and all demos under networking (writing so no1 covers the same demos). If I fail to convert any, I will say. By the end of next weekend I should be done, the conversion is easy in most.

**Comment by TheYellowArchitect** [comment]
[pong is broken](https://files.catbox.moe/dz58ng.mkv)

**Comment by TheYellowArchitect** [comment]
### Viewport
- 2D in 3D is perfect as is, I doubt it even asked for conversion from 4.0 to 4.1
- 3D in 2D works as expected

**Comment by TheYellowArchitect** [comment]
### Audio Tutorials
All work great, except the text-to-speech. I cannot test it at all because I guess it uses default OS voicecharacters or whatever they are called and I have none.
![weird-error](https://github.com/godotengine/godot-demo-projects/assets/25630803/c33d7c71-7ede-4b7a-924d-6595a64183ad)

 All others work as expected, some surprised me tbh. If I had seen them on a new godotengine post on a new audio feature I would be hyped, yet it's already a thing... Some examples:

- MIDI Piano can be stretched in a week to allow the player character to play piano for fun (cool easter egg which would take months on a custom game engine, and that's why no games do it, e.g. background piano being playable)
- Spectrum for custom UI stuff (scifi games)
- Text-To-Speech for voice-acting (though there are AI tools nowadays to get easy high quality voiceacting)
- generator and device_changer are neat for changing the soundscape for easter eggs.
- mic recorder, cant see it usable in singleplayer games, but still great feature.

**Comment by TheYellowArchitect** [comment]
### GUI Tutorials

- GDPaint works as expected.
- Drag_and_drop works as expected.

**Comment by Calinou** [comment]
I went through all demos with 4.2.dev4 and ticked off which ones work correctly and which ones remain to be ported.

**Comment by Alex2782** [comment]
`Accelerometer, Magnetometer and Gyro Demo`
has the status `the demo still needs to be ported manually.`

is it this example?
https://github.com/godotengine/godot-demo-projects/tree/master/mobile/sensors

2 weeks ago I tried on my android tablet and found no problems, what changes exactly are still needed there?

**Comment by Alex2782** [comment]
@Calinou: Has anyone ported `TileMap` to version 4 yet?
![image](https://github.com/godotengine/godot-demo-projects/assets/41921395/264a8240-06fc-43e2-b7c2-69d816b493c9)

![image](https://github.com/godotengine/godot-demo-projects/assets/41921395/cd01b077-9ea6-4fa6-be69-d6f3ac516539)

I have ported some scripts. But how `Grid.gd (extends TileMap)` would have to be ported, I unfortunately do not know yet.  (some lines commented out)

Should I still create PR? So that others can continue. I would have to get better acquainted with TileMap first.

https://github.com/godotengine/godot-demo-projects/compare/master...Alex2782:godot-demo-projects:role_playing_game_v4.1


**Current state**:
![image](https://github.com/godotengine/godot-demo-projects/assets/41921395/24d11083-9569-41b7-8434-707039ed4ccf)

**Comment by Alex2782** [comment]
(3D) [kinematic_character](https://github.com/godotengine/godot-demo-projects/tree/master/3d/kinematic_character)

The description would still have to be adapted here. The 2 classes no longer exist. The name 'Kinematic Character 3D' would have to be renamed to 'Character Body 3D'?

> This demo uses a [`KinematicBody`](https://docs.godotengine.org/en/latest/classes/class_kinematicbody.html)
> for the player, and [`StaticBody`](https://docs.godotengine.org/en/latest/classes/class_staticbody.html) for the level. 

... the player `extends CharacterBody3D` and the level `?`

**Comment by Alex2782** [comment]
> I've tested the demo in 4.2.dev4 and nothing happens when I press various buttons. This will need a second look by @bruvzg who initially made this demo.

would work again with this adjustment
https://github.com/godotengine/godot/pull/82859

**Comment by Alex2782** [comment]
`Pong Multiplayer` no problems detected

v4.1.2.stable.official [399c9dc39]

**Comment by AliceGrey** [comment]
@aaronfranke `2.5D Demo with GDScript` is not fully converted to 4. The node25d plugin is broken and does not currently run on 4.

**Comment by Alex2782** [comment]
> @aaronfranke `2.5D Demo with GDScript` is not fully converted to 4. The node25d plugin is broken and does not currently run on 4.

I could not find any problems. Master Branch was updated 5 months ago, version 4.0.
I tested with Godot Version 4.2 Beta4

https://github.com/godotengine/godot-demo-projects/tree/master/misc/2.5d

**Comment by MatrixFrog** [comment]
@Alex2782 I don't understand TileMap extremely well yet either but I think I'm making progress on that 2D RPG demo, using the branch you posted above as a starting point: https://github.com/godotengine/godot-demo-projects/pull/1009

Will try to get it fully working in the next few days...

**Comment by GeorgeS2019** [comment]
We could consider demonstrating the C# examples using [Godot 4 ](https://github.com/godotengine/godot-proposals/issues/432#issuecomment-1987336025)(Scene) **Test-Driven-Development** (**TDD**) in e.g. VS2022

https://youtu.be/JLI0cIYMgr0

using the [latest](https://github.com/MikeSchulze/gdUnit4Mono/pull/37) gdUnit4API **TestAdaptor** from @MikeSchulze 

https://github.com/godotengine/godot-proposals/assets/49812372/72bb91a1-dd1c-4467-8d07-3b0d3beadd4b

**Comment by tinybitofheaven** [comment]
I wanted to ask if I could work on Multiplayer Pong with C#? I've already started looking into converting it, it seems that multiplayer in Godot 4 is radically different from Godot 3 and a lot of RPC syntax is hard to find a direction conversions in C#. I'll keep working on it since I don't think anyone else is right now, but let me know if I'm wrong about that.

Also wanted to mention that one of the links in the tracker post is linked incorrectly.
![Screenshot 2024-04-08 205410](https://github.com/godotengine/godot-demo-projects/assets/32124445/1403de8d-4b8a-4735-aaf8-2c8857daf904)

**Comment by Calinou** [comment]
> I wanted to ask if I could work on Multiplayer Pong with C#? I've already started looking into converting it, it seems that multiplayer in Godot 4 is radically different from Godot 3 and a lot of RPC syntax is hard to find a direction conversions in C#. I'll keep working on it since I don't think anyone else is right now, but let me know if I'm wrong about that.

Yes, feel free to open a PR to update it :slightly_smiling_face: 

> Also wanted to mention that one of the links in the tracker post is linked incorrectly.

Fixed.

**Comment by tinybitofheaven** [comment]
I believe I've gotten Multiplayer Pong C# working in Godot 4.2.
However, I wanted to follow up on something. It seems that there isn't much documentation on implementing Multiplayer with C# in Godot 4, at least I couldn't find it very easily. In the [high-level multiplayer](https://docs.godotengine.org/en/stable/tutorials/networking/high_level_multiplayer.html) documentation, there's only GDScript. Is there plans to add more C# multiplayer documentation in the future? Or does the documentation exist and I just missed it?

**Comment by GeorgeS2019** [comment]
3D IK is coming 

https://github.com/godotengine/godot-demo-projects/pull/1271
